File: agenkit-scala/src/main/scala/io/agenkit/composition/SequentialAgent.scala
package io.agenkit.composition

import io.agenkit.core.*
import scala.concurrent.{ExecutionContext, Future}
import java.util.concurrent.atomic.AtomicLong

/** Pipes the output of each stage as input to the next stage. */
class SequentialAgent(
  val name: String,
  stages: List[Agent]
) extends Agent:
  private val _processedCount = new AtomicLong(0)

  def capabilities: List[String] = List("sequential", "pipeline")

  def process(message: Message)(using ExecutionContext): Future[Message] =
    _processedCount.incrementAndGet()
    stages.foldLeft(Future.successful(message)) { (msgFut, stage) =>
      msgFut.flatMap(stage.process)
    }

  def introspect(): IntrospectionResult =
    IntrospectionResult(
      name = name,
      metadata = Map("stageCount" -> stages.size, "stages" -> stages.map(_.name)),
      capabilities = capabilities,
      processedMessages = _processedCount.get()
    )

File: agenkit-java/src/main/java/io/agenkit/composition/SequentialAgent.java
package io.agenkit.composition;

import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * Composes agents into a sequential pipeline.
 * The output of each agent becomes the input to the next.
 */
public final class SequentialAgent implements Agent {

    private final String name;
    private final List<Agent> pipeline;

    public SequentialAgent(String name, List<Agent> pipeline) {
        this.name = name;
        this.pipeline = List.copyOf(pipeline);
    }

    @Override
    public String getName() { return name; }

    @Override
    public List<String> getCapabilities() {
        return List.of("sequential_composition", "pipeline");
    }

    @Override
    public CompletableFuture<Message> process(Message message) {
        CompletableFuture<Message> chain = CompletableFuture.completedFuture(message);
        for (Agent agent : pipeline) {
            chain = chain.thenCompose(agent::process);
        }
        return chain.thenApply(result ->
                result.withMetadata("pipeline_length", pipeline.size()));
    }

    @Override
    public IntrospectionResult introspect() {
        List<String> agentNames = pipeline.stream().map(Agent::getName).toList();
        Map<String, Object> state = new HashMap<>();
        state.put("pipelineLength", pipeline.size());
        return new IntrospectionResult(name, getCapabilities(), null, state, agentNames);
    }
}
